### PR TITLE
Sage-Select

### LINKED ISSUE #198: Sage Form Select

### PR DESCRIPTION
**Pull request by AndrwM**
Resolves: #198

# _Sage Select_ 🔮

- Adds sage-select element
- Fixes `position()` mixin (I broke it in my last PR)

### Notes
`<select>` doesn't have an empty/placeholder pseudo selector so I'm using `:invalid` to adjust the positioning of the label.

Feedback on the variable structure requested. The structure in this PR is more in-line with a "private variable" pattern. These variable will become fully "private" once we start using `@use`.

### Preview

screencast 2020-04-28 09-55-57

### CODE REVIEW

File: app/assets/stylesheets/sage/system/core/_mixins.scss [hunk 1]
```diff
@@ -132,7 +132,7 @@
   } @else if $value != unset {
     @return $value;
   } @else {
-    @return false;
+    @return null;
```
Comment (philschanely) [inline app/assets/stylesheets/sage/system/core/_mixins.scss:135]: 😍 Good catch here.

File: app/assets/stylesheets/sage/system/patterns/elements/_form_select.scss [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+.sage-select__label {
+  position: absolute;
+  white-space: nowrap;
+  pointer-events: none;
+  line-height: initial;
+
+  @include position((($-sage-selectfield-height / 2) - $sage-body-font-size), unset, unset, $-sage-selectfield-padding-x);
+}
+
+%sage-select__label--active {
```
Comment (philschanely) [inline app/assets/stylesheets/sage/system/patterns/elements/_form_select.scss:41]: Is this modularization necessary? Should we perhaps have a centralized label mixin? Am I missing something?

File: app/assets/stylesheets/sage/system/patterns/elements/_form_select.scss [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  color: transparent;
+  border: $-sage-selectfield-border-width solid $-sage-selectfield-border-color;
+  border-radius: $-sage-selectfield-border-radius;
+  background: transparent;
+  outline: none;
+  cursor: pointer;
+
+  // The <Select> is set as [required] & therefore is invalid when the value
+  // is empty, we're mapping :invalid to "empty" for more legible code.
+  $empty: "invalid";
```
Comment (philschanely) [inline app/assets/stylesheets/sage/system/patterns/elements/_form_select.scss:86]: 🥇

**Review by philschanely — APPROVED** [review]
Nice work. Just a non-blocking question.

File: app/assets/stylesheets/sage/system/patterns/elements/_form_select.scss [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  line-height: initial;
+
+  @include position((($-sage-selectfield-height / 2) - $sage-body-font-size), unset, unset, $-sage-selectfield-padding-x);
+}
+
+%sage-select__label--active {
+  @extend %t-sage-body-xsmall-semi;
+
+  padding: 0 $-sage-selectfield-padding-label;
+  color: inherit;
```
Comment (teenwolfblitzer) [inline app/assets/stylesheets/sage/system/patterns/elements/_form_select.scss]: Was the duped value here intentional? Looks like stylelint is choking on this one.

File: app/assets/stylesheets/sage/system/patterns/elements/_form_select.scss (hunk 2 above)
Comment (AndrwM) [reply app/assets/stylesheets/sage/system/patterns/elements/_form_select.scss:41]: Yeah, I think we should. Good idea. Right now we're using this style within:
- textarea
- dropdown
- string input
- select

There's some variation in how the positioning needs to be handled in some of the components but across the board it's the same visual style.

Comment (AndrwM) [reply app/assets/stylesheets/sage/system/patterns/elements/_form_select.scss:41]: The verdict here: Shipping this as-is and refactoring in a followup PR.

File: app/assets/stylesheets/sage/system/patterns/elements/_form_select.scss [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  border-radius: $-sage-selectfield-border-radius;
+  background: transparent;
+  outline: none;
+  cursor: pointer;
+
+  // The <Select> is set as [required] & therefore is invalid when the value
+  // is empty, we're mapping :invalid to "empty" for more legible code.
+  $empty: "invalid";
+
+  &:hover:#{$empty} {
```
Comment (teenwolfblitzer) [inline app/assets/stylesheets/sage/system/patterns/elements/_form_select.scss:89]: Could we also add on `:focus` to this (and the style below?)

**Review by teenwolfblitzer — COMMENTED** [review]
Couple of small items to check out, looking great so far!

I did run into a weird bug with Firefox where the dropdown text becomes invisible. It didn't happen consistently, so it could just be my system.:
|  default | error  |
|--------|--------|
|ff-select|ff-select2|

File: app/assets/stylesheets/sage/system/patterns/elements/_form_select.scss (hunk 4 above)
Comment (AndrwM) [reply app/assets/stylesheets/sage/system/patterns/elements/_form_select.scss]: Oops, nope! Thanks for catching

**Comment by AndrwM** [comment]
@monkeypox8 Was able to isolate that bug. TIL, firefox allows you to set the color of `<option>`. It was inheriting `transparent` from `<select>`.

Addressed this and the other issues above. Thanks!

**Review by teenwolfblitzer — APPROVED** [review]
👍 🔥 Looks great!

**Pull request merged**